DOC: Add examples for MultiIndex.get_locs + cleanups

topper-123 opened the pull request:
- [x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Adds examples to ``MultiIndex.get_locs`` and also some changes to ``.get_loc`` and ``.get_loc_values``.

Commit:
Add examples for MultiIndex.get_locs + cleanups

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1982,18 +1982,23 @@ def _partial_tup_index(self, tup, side='left'):
-        boolean mask array, otherwise it is always a slice or int.
+        Get location for a label or a tuple of labels as an integer, slice or
+        boolean mask.
+
```
I would put this Note in Notes.

Commit:
clarifies input to get_loc is sequence

topper-123 commented:
I've made some changes and put them in a new commit as it contains some (very minor) corrections to previous commits also.

#17655 had the reference to ``api.types.CategoricalDType`` wrong (``pandas.`` was prepended which made the link unusable). Given @TomAugspurger excellent new commit on CategoricalDType, I've changed the location + added a small text description.

#17653  had too many back-ticks in the links in section "Examples". I've removed those.

jreback commented on pandas/core/categorical.py:
```diff
@@ -223,7 +223,7 @@ class Categorical(PandasObject):
 
     See also
     --------
-    pandas.api.types.CategoricalDtype
```
does this property xref here? (CategoricalDtype is NOT in the main namespace)

topper-123 replied:
Ok, I've changed it to ``api.types.CategoricalDtype``. I had misunderstood the location.

topper-123 replied:
Looking more into it, I actually think ``CategoricalDtype`` isn't shown at all in the generated API docs, so the reference will be not work in ant case. Is this correct?

jorisvandenbossche replied:
Yeah, the problem is not that 'pandas' is there (that is perfectly OK, and in this case I even prefer to show the full path for clarity, as `api` is not a commonly used submodule of pandas), the issue is that this page does not exist, so it cannot link to it

topper-123 replied:
I'll put ``pandas.`` back in then.

jorisvandenbossche replied:
Yep. You can also add it to the api.rst pages so the link will work

topper-123 replied:
Change uploaded.

There seems to be something in the API.rst (line 649), but it mustn't work. @TomAugspurger has started looking into it, I'll wait till he figures it out. 

If I get to know what I should add to API.rst, I can add it to this PR.

TomAugspurger replied:
In api.rst it should be `pandas.api.type.CategoricalDtype` (missing the pandas).

But the autoclass fit the best with the rest of the autosummary's, so I may change it. For now, I'd just change your reference to `pandas.api.type.CategoricalDtype`, and I'll make sure they're valid before the release.

TomAugspurger commented:
Seems like some issue with the API reference. I'm looking into it.

On Tue, Sep 26, 2017 at 6:48 AM, topper-123 <notifications@github.com>
wrote:

> *@topper-123* commented on this pull request.
> ------------------------------
>
> In pandas/core/categorical.py
> <https://github.com/pandas-dev/pandas/pull/17675#discussion_r141031952>:
>
> > @@ -223,7 +223,7 @@ class Categorical(PandasObject):
>
>      See also
>      --------
> -    pandas.api.types.CategoricalDtype
>
> Looking more into it, I actually think CategoricalDtype isn't shown at
> all in the generated API docs, so the reference will be not work in ant
> case. Is this correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/17675#discussion_r141031952>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIj7WrfR-Cpu1g9U1bQrEdH6WLX7fks5smOShgaJpZM4Pjbs9>
> .
>

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -1982,33 +1982,41 @@ def _partial_tup_index(self, tup, side='left'):
-        get_locs : Given a tuple of slices/lists/labels/boolean indexer to a
-                   level-wise spec, produce an indexer to extract those
-                   locations.
+        get_locs : Get location for a label/slice/list/mask or a sequence of
```
get_locs -> MultiIndex.get_locs

topper-123 replied:
Ok

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -2328,23 +2337,40 @@ def convert_indexer(start, stop, step, indexer=indexer, labels=labels):
         Parameters
         ----------
-        key : tuple of (slices/list/labels)
+        seq : label/slice/list/mask or a sequence of such.
```
It might be good to explain how the 'sequence of such' should be interpreted

(I should check the code myself to actually know it, but are the different elements of the list mapped to the different levels?)

jorisvandenbossche replied:
maybe add here "(one for each level)" as you did in get_loc

topper-123 replied:
I've made a new proposal. This is very difficult to explain in short sentences, I find.

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -2328,23 +2337,40 @@ def convert_indexer(start, stop, step, indexer=indexer, labels=labels):
         Returns
         -------
-        locs : integer list of locations or boolean indexer suitable
+        locs : integer, slice or boolean mask suitable
```
I don't think this is correct: you changed 'integer list of locations' to 'integer', but from the examples it seems to return an array of integers, so the 'list' was more correct

topper-123 replied:
Well, tup indicates that this method takes a tuple, but it also takes other sequences like lists. In an example I tried it actually choked on tuple, because it thought the tuple was the label of a level. My experience from the last days it therefore is that it's probaby best to use lists rather than tuples (but tuple may work, depending on context).

On the other hand get_loc (without s) demands a tuple and chokes on lists, so it's not so easy to remember what is accepted where.

I can say "list or such", to discourage using tuples?

jorisvandenbossche replied:
Note that my comment was about the *return value*, not the input (and there indeed tuples vs lists sometimes are slightly different but not always, which can be confusing)

topper-123 replied:
Ok, changed

topper-123 replied:
Note, I actually couldn't manage to get it to return a boolean mask.

Could you confirm that? I haven't changed that part, but if it's not true, it'd be better it didn't say that...

jorisvandenbossche replied:
Yes, looking at the code, I actually think it can only return an integer array. That would mean that the explanation is just wrong.

But I am not too familiar with this. (@jreback ?)

Commit:
Small addition/cleanup + resolved merge conflict

topper-123 commented:
Is there anything I should do here wrt. the CircleCI tests failing? The test failure doesn't seem to related to my PR.

jorisvandenbossche commented:
No, you don't need to care, there are some problems in general: https://github.com/pandas-dev/pandas/issues/17702

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -2328,23 +2337,40 @@ def convert_indexer(start, stop, step, indexer=indexer, labels=labels):
+        >>> mi = pd.MultiIndex.from_arrays([list('abb'), list('def')])
+
+        >>> mi.get_locs('b')
+        slice(1, 3, None)
```
If I run this locally, I actually get an array as result:

```
In [109]: mi = pd.MultiIndex.from_arrays([list('abb'), list('def')])

In [110]: mi.get_locs('b')
Out[110]: array([1, 2])

```

topper-123 replied:
Yeah, me too, actually. It seems the func then only return arrays?

topper-123 commented:
I've looked some more at the return value. The return is always ``index._values`` and ``index`` is always numerical. So the method does indeed only return integer arrays. I've changed the PR correspondingly.

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -1982,33 +1982,41 @@ def _partial_tup_index(self, tup, side='left'):
+        Notes
+        ------
+        The key cannot be a slice, list of same-level labels, a boolean mask,
+        or a sequence of such. If you want to use those, use :meth:`get_locs`
```
get_locs -> MultiIndex.get_locs  (I think, it would actually be interesting to know whether it works like this if it is a method of the same class. Did you build it locally and checked? But I think this will not work, because in the generated rst docstring pages, a `.. currentmodule:: pandas` is included)

topper-123 replied:
I've Changed it.

It it possible to to build just the API docs or maybe a subgroup thereof? Working with doc is a but cumbersome, when building it takes so much time.

jreback replied:
you pass the arg — single api

jorisvandenbossche replied:
Yes, it can indeed be very cumbersome. It's a very long time ago that I tried the `--single` flag. It was also meant to eg run only a single .rst file. But I can remember it also didn't work smoothly. 

If you think of other ways to make this easier, very welcome! (I once had a script to build only a single docstring, I could try to look that up, but not sure that such links then would work anyhow)

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -2328,23 +2337,41 @@ def convert_indexer(start, stop, step, indexer=indexer, labels=labels):
-        key : tuple of (slices/list/labels)
+        seq : label/slice/list/mask or a sequence of such (one for each used
+              level. If a level should not be used, set it to
+              ``slice(None)``).
```
Multi-line type specifications don't really work (does not render correctly in sphinx, unless you use `\` to indicate line continuation). But you can also spit this up:

```
seq : label/slice/list/mask or a sequence of such (one for each level)
    More explanation ...
```

topper-123 replied:
changed

jorisvandenbossche reviewed (commented):
Some final minor comments!

jorisvandenbossche commented:
@topper-123 Thanks a lot!
